hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java

```diff
@@ -507,6 +508,10 @@ public StakingInfo stakingInfo(final MerkleAccount account) {
 
 		if (account.getStakePeriodStart() > 0) {
 			stakingInfo.setStakePeriodStart(Timestamp.newBuilder().setSeconds(account.getStakePeriodStart()).build());
+			if (account.mayHavePendingReward()) {
+				rewardCalculator.computePendingRewards(account);
```

Comment:
⚠️ The `computePendingRewards` method is not thread-safe (i.e., this changes fields in the singleton `rewardCalculator` _also_ being used in `handleTransaction` and may cause an ISS). Even worse, `computePendingRewards` mutates its `account` argument, so this will corrupt the state of any node that answers a `getAccountInfo` query for an account with pending rewards. 😰 

Because of this kind of risk, we should try write `StateView` methods in terms of **only** the `stateChildren` field.

In this case, unfortunately, the `stateChildren` do not include the current staking period or the first effective staking period. A fine option for now would be to add a thread-safe method to the `RewardCalculator`,

```
public long estimatePendingRewards(final MerkleAccount account, final MerkleStakingInfo stakedNodeInfo) {
    return computeRewardFromDetails(
        account,
        stakedNodeInfo,
        stakePeriodManager.currentStakePeriod(),
        stakePeriodManager.effectivePeriod(account.getStakePeriodStart()));
}
```
...where we have added a `computeRewardFromDetails` that will also be used by [`computeReward`](https://github.com/hashgraph/hedera-services/blob/9a04d22d5087a5e766ac3c0379c73040176ac5c8/hedera-node/src/main/java/com/hedera/services/ledger/accounts/staking/RewardCalculator.java#L77)

Reply:
I agree with `computeRewardFromDetails `. But, regarding the comment `computePendingReward` doesn't set any fields on account. If we had reset the `accountReward` and `accountUpdatedStakePeriodStart` in addition to `rewardsPaid` [here](https://github.com/hashgraph/hedera-services/blob/3051-M-finalizeSideEffects/hedera-node/src/main/java/com/hedera/services/ledger/accounts/staking/RewardCalculator.java#L89), would that have solved the issue ? Since those fields will be reset every handle